SPHINXsys/src/shared/materials/diffusion_reaction.h

```diff
@@ -163,93 +163,128 @@ namespace SPH
 	class BaseReactionModel
 	{
 	public:
+		const static int NumSpecies = NUM_SPECIES;
 		typedef std::array<Real, NUM_SPECIES> LocalSpecies;
 		typedef std::array<std::string, NUM_SPECIES> SpeciesNames;
 		typedef std::function<Real(LocalSpecies &)> ReactionFunctor;
-		IndexVector reactive_species_;
 		StdVec<ReactionFunctor> get_production_rates_;
 		StdVec<ReactionFunctor> get_loss_rates_;
 
-		explicit BaseReactionModel(SpeciesNames species_name_list)
-			: reaction_model_("BaseReactionModel"), species_name_list_(species_name_list)
+		BaseReactionModel()
 		{
-			for (size_t i = 0; i != species_name_list.size(); ++i)
+			std::cout << "\n Error: default constructor for non-empty reaction model!" << std::endl;
+			std::cout << __FILE__ << ':' << __LINE__ << std::endl;
+			exit(1);
+		};
+
+		explicit BaseReactionModel(const SpeciesNames &species_names)
+			: species_names_(species_names)
+		{
+			if (NUM_SPECIES == 0)
+			{
+				reaction_model_ = "EmptyReactionModel";
+			}
+			else
 			{
-				species_indexes_map_.insert(make_pair(species_name_list[i], i));
+				reaction_model_ = "BaseReactionModel";
+				for (size_t i = 0; i != species_names.size(); ++i)
+				{
+					species_indexes_map_.insert(make_pair(species_names[i], i));
+				}
 			}
 		};
 		virtual ~BaseReactionModel(){};
-		SpeciesNames getSpeciesNameList() { return species_name_list_; };
+		SpeciesNames &getSpeciesNames() { return species_names_; };
 
 	protected:
 		std::string reaction_model_;
-		SpeciesNames species_name_list_;
+		SpeciesNames species_names_;
 		std::map<std::string, size_t> species_indexes_map_;
 	};
+	/** explicit specialization for empty reaction model */
+	template <>
+	inline BaseReactionModel<0>::BaseReactionModel() : reaction_model_("EmptyReactionModel"){};
+	using NoReaction = BaseReactionModel<0>;
 
 	/**
 	 * @class DiffusionReaction
 	 * @brief Complex material for diffusion or/and reactions.
 	 */
-	template <class BaseMaterialType = BaseMaterial, int NUM_SPECIES = 1>
+	template <class BaseMaterialType = BaseMaterial, int NUM_REACTIVE_SPECIES = 0>
 	class DiffusionReaction : public BaseMaterialType
 	{
+	public:
+		const static int NumReactiveSpecies = NUM_REACTIVE_SPECIES;
```

Comment:
Does this make difference?

Reply:
It guarantees compile-time optimization and allows compile-time usage (e.g. template, if constexpr) of the variable. It is the preferred way for values known at compile-time. Besides the original version should have been `inline static const` rather than `static const` since it is defined in the class declaration.